transcode/transcode.go

```diff
@@ -570,47 +597,99 @@ func transcodeSegment(
 	duration := time.Since(start)
 	metrics.Metrics.TranscodeSegmentDurationSec.Observe(duration.Seconds())
 
-	for _, transcodedSegment := range tr.Renditions {
-		renditionIndex := getProfileIndex(transcodeProfiles, transcodedSegment.Name)
-		if renditionIndex == -1 {
-			return fmt.Errorf("failed to find profile with name %q while parsing rendition segment", transcodedSegment.Name)
+	err = processTranscodeResult(segment, transcodeRequest, sourceSegment, tr, encodedProfiles, targetOSURL, transcodedStats, renditionList, segmentChannel)
+	if err != nil {
+		return fmt.Errorf("failed to process transcode result: %w", err)
+	}
+
+	return nil
+}
+
+// withPipedSource is used to duplicate the reading of the `in` reader in case we need a copy of the contents. If
+// `copySource` is false then the `in` reader is returned as is. Otherwise, then a non-nill buffer will be returned and
+// filled after the returned reader is consumed (if present). If no reader is returned (empty transcodeProfiles) the
+// buffer will be already filled with the contents of the `in` reader.
+func withPipedSource(in io.Reader, copySource bool, transcodeProfiles []video.EncodedProfile) (io.Reader, *bytes.Buffer, error) {
+	if !copySource {
+		return in, nil, nil
+	}
+
+	source := new(bytes.Buffer)
+	if len(transcodeProfiles) == 0 {
+		// This is a copy-only job, so skip transcoding with the broadcaster
+		_, err := io.Copy(source, in)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to copy source segment: %s", err)
+		}
+		return nil, source, nil
+	}
+
+	// Otherwise if there are profiles to transcode, then tee the source segment to the buffer
+	return io.TeeReader(in, source), source, nil
+}
+
+func processTranscodeResult(
+	segment segmentInfo,
+	transcodeRequest TranscodeSegmentRequest,
+	sourceSegment *bytes.Buffer,
+	transcodeResult clients.TranscodeResult,
+	encodedProfiles []video.EncodedProfile,
+	targetOSURL *url.URL,
+	transcodedStats []*video.RenditionStats,
+	renditionList *video.TRenditionList,
+	segmentChannel chan<- video.TranscodedSegmentInfo) error {
+
+	for renditionIndex, profile := range encodedProfiles {
+		var mediaData []byte
+		if profile.Copy {
+			mediaData = sourceSegment.Bytes()
+		} else {
+			for _, transcodedSegment := range transcodeResult.Renditions {
+				if transcodedSegment.Name == profile.Name {
+					mediaData = transcodedSegment.MediaData
+					break
+				}
+			}
+		}
+		if mediaData == nil {
+			return fmt.Errorf("failed to find rendition with name %q while parsing transcode result", profile.Name)
 		}
 
-		targetRenditionURL, err := url.JoinPath(targetOSURL.String(), transcodedSegment.Name)
+		targetRenditionURL, err := url.JoinPath(targetOSURL.String(), profile.Name)
 		if err != nil {
 			return fmt.Errorf("error building rendition segment URL %q: %s", targetRenditionURL, err)
 		}
 
 		if transcodeRequest.GenerateMP4 {
```

Comment:
So we definitely want to be generating MP4s for fishtanks long recordings?

Reply:
Yeah the logic should stay the same here! Generate the mp4s if they are shorter than our threshold (I think 6h? or 12h)